From be6c4b5193ea339c44e0e3c88ae4ef26e9dfcb3d Mon Sep 17 00:00:00 2001
From: Nick Xie <nick@khadas.com>
Date: Tue, 20 Jan 2026 11:24:12 +0800
Subject: [PATCH 66/99] FROMLIST(v1): arm64: dts: amlogic: meson-s4: add UART_A
 node

Add the UART_A node and its related pinctrl definitions to the
Meson S4 SoC dtsi.

This interface is typically used for Bluetooth communication on
boards like the Khadas VIM1S.

Signed-off-by: Nick Xie <nick@khadas.com>
---
 arch/arm64/boot/dts/amlogic/meson-s4.dtsi | 22 ++++++++++++++++++++++
 1 file changed, 22 insertions(+)

diff --git a/arch/arm64/boot/dts/amlogic/meson-s4.dtsi b/arch/arm64/boot/dts/amlogic/meson-s4.dtsi
index dfc0a30a6e61..b1d66054a6e8 100644
--- a/arch/arm64/boot/dts/amlogic/meson-s4.dtsi
+++ b/arch/arm64/boot/dts/amlogic/meson-s4.dtsi
@@ -163,6 +163,16 @@ mux {
 					};
 				};
 
+				a_uart_pins: a_uart {
+					mux {
+						groups = "uart_a_tx",
+						       "uart_a_rx",
+						       "uart_a_cts",
+						       "uart_a_rts";
+						function = "uart_a";
+					};
+				};
+
 				i2c0_pins1: i2c0-pins1 {
 					mux {
 						groups = "i2c0_sda",
@@ -814,6 +824,18 @@ mdio0: mdio {
 			};
 		};
 
+		uart_a: serial@fe078000 {
+			compatible = "amlogic,meson-s4-uart",
+				     "amlogic,meson-ao-uart";
+			reg = <0x0 0xfe078000 0x0 0x18>;
+			interrupts = <GIC_SPI 168 IRQ_TYPE_EDGE_RISING>;
+			clocks = <&xtal>, <&clkc_periphs CLKID_UART_A>, <&xtal>;
+			clock-names = "xtal", "pclk", "baud";
+			pinctrl-names = "default";
+			pinctrl-0 = <&a_uart_pins>;
+			status = "disabled";
+		};
+
 		sdio: mmc@fe088000 {
 			compatible = "amlogic,meson-axg-mmc";
 			reg = <0x0 0xfe088000 0x0 0x800>;
-- 
2.43.0

